core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -976,6 +975,231 @@ class ReplicaManager(val config: KafkaConfig,
     }
   }
 
+  /**
+   * Append messages to offsets topic, and wait for them to be replicated to other replicas;
+   * the callback function will be triggered either when timeout or the required acks are satisfied;
+   * if the callback function itself is already synchronized on some object then pass this object to avoid deadlock.
+   *
+   * Noted that all pending delayed check operations are stored in a queue. All callers to ReplicaManager.appendRecords()
+   * are expected to call ActionQueue.tryCompleteActions for all affected partitions, without holding any conflicting
+   * locks.
+   *
+   * @param timeout                       maximum time we will wait to append before returning
+   * @param requiredAcks                  number of replicas who must acknowledge the append before sending the response
+   * @param internalTopicsAllowed         boolean indicating whether internal topics can be appended to
+   * @param origin                        source of the append request (ie, client, replication, coordinator)
+   * @param entriesPerPartition           the records per partition to be appended
+   * @param responseCallback              callback for sending the response
+   * @param delayedProduceLock            lock for the delayed actions
+   * @param recordValidationStatsCallback callback for updating stats on record conversions
+   * @param requestLocal                  container for the stateful instances scoped to this request -- this must correspond to the
+   *                                      thread calling this method
+   * @param actionQueue                   the action queue to use. ReplicaManager#defaultActionQueue is used by default.
+   * @param verificationGuards            the mapping from topic partition to verification guards if transaction verification is used
+   * @param preAppendErrors               the mapping from topic partition to LogAppendResult for errors that occurred before appending
+   */
+  def appendForGroup(timeout: Long,
```

Comment:
Can we generalize the name? I don't see any logic specific to offsets or groups.

Reply:
I planned to take this out after the refactor. It is only used for appendForGroup to minimize the diff.

Reply:
Makes sense. But any harm having a more general name for now? It might be a week or two before the refactor gets checked in.

Reply:
I guess I just didn't see anything else using it before I refactored and wanted to make the usage clear. 

I didn't want to name it appendRecords as to not cause conflicts with that flow. What name were you thinking?

Reply:
We should also add a comment that would reflect 2 points:

1. (For the maintainer of this code) -- this code must not return until the local write is done, it is an important invariant that the callers rely upon.  Otherwise it looks like a generic async call that can return and continue asynchronously at any point.  This way, if an additional async stage is required in this function before the  the local write is complete, the maintainer would know to hunt down all usages of this function and figure out the correct action.
2. (For the caller of this code) -- a quick example of the full workflow of how the caller should use this method: call maybeStartTransactionVerificationForPartition with a callback that would call this method.

Reply:
Adding comment in the refactor should be fine.

Reply:
If we keep the name, how about we add a check to ensure that the write is for __consumer_offsets? Also, we can drop the `internalTopicsAllowed` and `appendOrigin` arguments since they will be implicit.